substrate/client/rpc-spec-v2/src/archive/tests.rs

```diff
@@ -838,3 +852,256 @@ async fn archive_storage_discarded_items() {
 		_ => panic!("Unexpected result"),
 	};
 }
+
+#[tokio::test]
+async fn archive_storage_diff_main_trie() {
+	let (client, api) = setup_api(MAX_PAGINATION_LIMIT, MAX_QUERIED_LIMIT);
+
+	let mut builder = BlockBuilderBuilder::new(&*client)
+		.on_parent_block(client.chain_info().genesis_hash)
+		.with_parent_block_number(0)
+		.build()
+		.unwrap();
+	builder.push_storage_change(b":A".to_vec(), Some(b"B".to_vec())).unwrap();
+	builder.push_storage_change(b":AA".to_vec(), Some(b"BB".to_vec())).unwrap();
+	let prev_block = builder.build().unwrap().block;
+	let prev_hash = format!("{:?}", prev_block.header.hash());
+	client.import(BlockOrigin::Own, prev_block.clone()).await.unwrap();
+
+	let mut builder = BlockBuilderBuilder::new(&*client)
+		.on_parent_block(prev_block.hash())
+		.with_parent_block_number(1)
+		.build()
+		.unwrap();
+	builder.push_storage_change(b":A".to_vec(), Some(b"11".to_vec())).unwrap();
+	builder.push_storage_change(b":AA".to_vec(), Some(b"22".to_vec())).unwrap();
+	builder.push_storage_change(b":AAA".to_vec(), Some(b"222".to_vec())).unwrap();
+	let block = builder.build().unwrap().block;
+	let block_hash = format!("{:?}", block.header.hash());
+	client.import(BlockOrigin::Own, block.clone()).await.unwrap();
+
+	// Search for items in the main trie:
+	// - values of keys under ":A"
+	// - hashes of keys under ":AA"
+	let items = vec![
+		ArchiveStorageDiffItem::<String> {
+			key: hex_string(b":A"),
+			return_type: ArchiveStorageDiffType::Value,
+			child_trie_key: None,
+		},
+		ArchiveStorageDiffItem::<String> {
+			key: hex_string(b":AA"),
+			return_type: ArchiveStorageDiffType::Hash,
+			child_trie_key: None,
+		},
+	];
+	let mut sub = api
+		.subscribe_unbounded(
+			"archive_unstable_storageDiff",
+			rpc_params![&block_hash, items.clone(), &prev_hash],
+		)
+		.await
+		.unwrap();
+
+	let event = get_next_event::<ArchiveStorageDiffEvent>(&mut sub).await;
+	assert_eq!(
+		ArchiveStorageDiffEvent::StorageDiff(ArchiveStorageDiffResult {
+			key: hex_string(b":A"),
+			result: StorageResultType::Value(hex_string(b"11")),
+			operation_type: ArchiveStorageDiffOperationType::Modified,
+			child_trie_key: None,
+		}),
+		event,
+	);
+
+	let event = get_next_event::<ArchiveStorageDiffEvent>(&mut sub).await;
+	assert_eq!(
+		ArchiveStorageDiffEvent::StorageDiff(ArchiveStorageDiffResult {
+			key: hex_string(b":AA"),
+			result: StorageResultType::Value(hex_string(b"22")),
+			operation_type: ArchiveStorageDiffOperationType::Modified,
+			child_trie_key: None,
+		}),
+		event,
+	);
```

Comment:
DQ: Shoudln't this one be deduplicated? ie it's handed back above. I had assumed that the query for A and AA would become just a query for A since AA is a child prefix of A but might be misunderstanding :)